librz/bin/format/pe/pe_relocs.c

```diff
@@ -0,0 +1,67 @@
+// SPDX-FileCopyrightText: 2024 Roee Toledano <roeetoledano10@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include "pe.h"
+
+bool PE_(bin_pe_has_base_relocs)(RZ_NONNULL RzBinPEObj *bin) {
+	rz_return_val_if_fail(bin, false);
+
+	return bin->relocs && (rz_vector_len(bin->relocs) > 0);
+}
+
+static bool read_reloc_ent_from_block(RZ_NONNULL RzVector /*<RzBinPeRelocEnt>*/ *relocs, RzBuffer *b, RzBinPeRelocBlock *block, ut64 *offset, const int big_endian) {
+	const ut32 reloc_block_end = *offset + block->block_size - 8; // block size includes the size of the next blocks entry, which is 8 bytes long
+	do {
+		RzBinPeRelocEnt reloc = { 0 };
+		if (!rz_buf_read_ble16_offset(b, offset, &reloc.raw_val, big_endian)) {
+			return false;
+		}
+		reloc.page_rva = block->page_rva;
+
+		rz_vector_push(relocs, &reloc);
+
+	} while (*offset < reloc_block_end);
+
+	return true;
+}
+
+static bool get_relocs_from_data_dir(RZ_NONNULL RzBinPEObj *bin, RZ_BORROW RZ_NONNULL RzVector /*<RzBinPeRelocEnt>*/ *relocs) {
+	RzBuffer *b = bin->b;
+	const st64 o_addr = rz_buf_tell(b);
+
+	// get offset in file of first reloc block
+	ut64 offset = PE_(bin_pe_rva_to_paddr)(bin, bin->nt_headers->optional_header.DataDirectory[PE_IMAGE_DIRECTORY_ENTRY_BASERELOC].VirtualAddress);
```

Comment:
Check first if `nt_headers` and `optional_header` aren't NULL. Also, make sure that the array access doesn't go out of bounds.

Reply:
Fixed. I just didn't add the bounds checking, since the `DataDirectory` array is always `PE_IMAGE_DIRECTORY_ENTRIES` length, so it's never out of bounds. I added a check to make sure the element at `PE_IMAGE_DIRECTORY_BASERELOC` is not null (even though it should never be null if relocs aren't stripped)